### DISCUSSION TITLE
(WIP) Improve the Camera Capture feature

### DISCUSSION THREAD

**Discussion by shun-iwasawa** [opentoonz/opentoonz#6786] [Roadmap]
This issue is for summarizing the feedback on the first release of Pencil Test feature (currently available only in the nightly build):
- [x] Add a maximize button to the window.
- [x] Enable Capture with shortcut key (enter key) - a temporary shortcut only for when the Pencil Test window is open.
- [x] Set the resolution to its highest value by default instead of its lowest.
- [x] Add option to save the images as they are captured.
- [x] Change the command name from Pencil Test to Camera Capture, in consideration of usage for stop motion animation.
- [x] The ability to shoot a reference video with camera in real-time for acting shots.
- [x] Timer / Time Lapse
- [x] BUG: The countdown number flips when the "Upside Down" check box is active.
- [x] Support the preferences option "Show ABC Appendix to the Frame Number in Xsheet Cell"
- [x] The ability to playback at least the captured images in the window.
- [ ] Multi frame onion skin.
- [x] The ability to load a reference image for lining up images between stop-motion sessions.
- [ ] BUG: Changing the resolution sometimes causes crash. 
- [ ] Support DSLR camera.
- [ ]  Add "No Camera" item to the camera combo box.
- [ ]  Add a "Mirror" option.
- [ ] Mac Stop Motion Support

**Comment by Rayek** [comment]
Love all of these suggestions.

**Comment by RodneyBaker** [comment]
Thanks from here also.
I'm really enjoying this new feature.

If I can be so bold as to suggest another enhancement it would be to allow
a timer for capture.  In this way the users hands would be freed to fully
concentrate on placing the next drawing (or in case of stopmotion... adjust
props/characters) in front of the camera.   I'd say the timespan might be a
10 count (or 10 seconds) between captures but this would optimally be a
number the user can set.  When the user is set to capture they hit
Capture/Start and off they go only pausing to wait for the image to be
captured.  Of course, some visual clue and/or sound would be nice to signal
the completion of the capture.  Then the user knows it's time to set the
next shot in anticipation of the next capture.

**Comment by ghost** [comment]
@RodneyBaker added your request to the list at top

**Comment by shun-iwasawa** [comment]
Updated the task list.

**Comment by ghost** [comment]
@shun-iwasawa Thank you so much for the improvements.  I added one more item to the list:
The ability to load a reference image for lining up images between stop-motion sessions.
This will help stop-motion artists continue where they left off from day to day.

**Comment by RodneyBaker** [comment]
Nice update!
Looking good on Win10.

I hope to purchase an external webcam as early as tomorrow and will test that also.
It works great with my internal webcam.

**Comment by shun-iwasawa** [comment]
Updated the task list. @RodneyBaker , thank you for reporting the bug.

>  Support the preferences option "Show ABC Appendix to the Frame Number in Xsheet Cell"

I think this task needs explanation. This demand arose from some Japanese animation studio.
Keyframe animators sometimes would like to insert a new drawing between already existing sequential frames. Let's say there is a level with 5 frames (1,2,3,4,5), when animator would like to insert a new drawing between frames 2 and 3.  In such case, rather than reframing following frames or inserting frame 6, they add to the new frame a modifier like "2A", "2′(prime symbol)" or "2.1". Allowing such exceptional frame number in Toonz was demanded in order to free users from converting frame numbers to sequential.
The preferences option "Show ABC Appendix to the Frame Number in Xsheet Cell" is my simple resolution to realize it. With this option, OpenToonz recognize the ones place digit of frame number as modifier and display in Xsheet with converting it to corresponding alphabet (e.g. 10 -> 1, 21->2A, 45->4E) so that users can add 9 (from "-A" to "-I") extra frames between two sequential integer frames afterwards. An advantage of this implementation is that it changes only UI and does not add any extra complexity to level's data structure.
The task added here is for reflecting this option to the UI of Camera Capture dialog as well.

**Comment by shun-iwasawa** [comment]
Bug fixed and updated the task list.

**Comment by RodneyBaker** [comment]
The Camera Test recognized my new Logitec c615 camera.  It appeared as an option after a reboot of my computer.  Looking good..

I am experiencing a few problems.

Crash after removing camera/software
After successful direct use of the Logitec camera I removed the MyCam software (although I might have to reinstall it if I decide to use my monitors internal camera).  Removing it caused a problem however in that the dropdown defaulted to the integrated camera rather than the Logitec.

The bigger problem here was that upon trying to relaunch OpenToonz startup failed because (I assume) it could not find the camera.   I tried to find a config file that would allow me to change the camera but couldn't so I uninstalled and reinstalled OpenToonz (needed the latest release anyway!)

This worked but when I attempted to change my camera in the dropdown OpenToonz crashed.

While I'm sure some code needs to be tweeked it seems to me that adding a default option to the dropdown menu of "No Camera" might be appropriate.  This would not only provide a means to 'park' all cameras but would put minds at ease with regard to security and not knowing whether the camera was on.  A user could set the camera to "No Camera" and press on.

I'll look to see if there is anything else I can find related to this issue.

That's the crash.  I'll add another post or two related to other minor suggestions for improvement.

As always thanks for everything you do!

**Update**:  I continue to have problems running Camera Capture because the dropdown is defaulting to my integrated camera.  Trying to select another camera from the dropdown menu crashses OpenToonz.  I'm reinstalling the ManyCam software to see if that will at least let me access the dropdown menu.  **Edit:  And it did.**

**Comment by RodneyBaker** [comment]
I note that initially the images captured by my webcams are upside down.
No problem OpenToonz Camera Capture can deal with that.

There is the upside down option for images captured from the camera but once inverted I see the image is flipped horizontally.   As there isn't a 'mirror' or 'mirror left/right' option we cannot get the images to show in their correct orientation.

Workarounds:  
There are at least two possible workarounds for this.  The first would be to check the camera software and see if it allows for the image to be reoriented.  Then when the images appear in OpenToonz no adjustment is necessary.  The second would be to capture the images in their present state and then adjust the orientation of the levels/frames.

Thanks!

**Comment by RodneyBaker** [comment]
Side note to @turtletooth 
The Camera Capture feature is opening in the portable release but no camera is selected and the dropdown menu appears not to allow for any selection.  Three dots (...) are all that appears at that menu location.  I have not yet been able to use the Camera Capture feature in the portable although I did open and then restart as you had identified that as a bug before.

**Comment by shun-iwasawa** [comment]
@RodneyBaker Thank you for reporting the problem and suggestions. I added your proposal to the list.

Regarding the "No Camera" item, I agree it is safe to set the initial state to "No Camera" considering the fact that some camera may cause crash of OpenToonz on connecting. But it will be troublesome if you have to choose camera every time you launch the Camera Capture. So I will make the sofware to save the last state of the camera selection in order to reproduce it on launch next time.

**Comment by RodneyBaker** [comment]
Thank you!

**Comment by RodneyBaker** [comment]
Here's an interface issue that should be able to be fixed easily...

I'm not sure where it first appeared but in current nightlly releases the button for 'Capture White BG' appears to not have an adequate height setting.  It might be that the button is set right but the area it is placed in isn't.
Here's a pic:
![buttonheight](https://cloud.githubusercontent.com/assets/2576432/17413400/ac68a74c-5a46-11e6-82d1-7c58f24d2ff2.jpg)

**Added/Update:  This bug in the height of the 'Capture White BG' button appears as the correct height if OpenToonz is full screen.**

**Comment by ghost** [comment]
Loading a reference image is taken care of in #751

**Comment by drodriguez848** [comment]
can you add support for canon eos utilities to be recognized by camera capture?

**Comment by shun-iwasawa** [comment]
@drodriguez848 
For now the camera capture feature relies on QCamera class in Qt library. Some DSLR cameras like canon eos are not recognized as camera devices.
FYI, by using [SparkoCam](http://sparkosoft.com/sparkocam) you can use Canon DSLR camera as a regular webcamera, I think it can be recognized by OT. SparkoCam offers a trial version ( watermark appears on the captured image ).

**Comment by ghost** [comment]
I downloaded the Canon sdk for dslr cameras.  This is how we would need to incorporate dslr into OT.

**Comment by artisteacher** [comment]
I have Canon SL2's to test with on the Mac end, when we get to that point.

**Comment by ghost** [comment]
@shun-iwasawa You around today?  I've got a question.  I'm trying to implement stop motion camera view right on the viewer instead of the popup.  I'm using onRasterImage() from the sceneviewer stage painter.  I have it showing the image, but I want to adjust its size.  Would I change the TAffine in the Stage::Player that I send to onRasterImage?

**Comment by shun-iwasawa** [comment]
@turtletooth 

> Would I change the TAffine in the Stage::Player that I send to onRasterImage?

Yes, I think so. I think adjusting the size of raster image can be done by the same way as adjusting the image dpi. That is to say, I think multiplying proper scale to `m_dpiAff` used in [here](https://github.com/opentoonz/opentoonz/blob/18a7c631a5941db6a89e34cb2aefe47e12953727/toonz/sources/toonzlib/stagevisitor.cpp#L954) would work.

**Comment by RodneyBaker** [comment]
The following are still listed as WIP for internal Camera Capture:
 * [x]  The ability to playback at least the captured images in the window.
 * [ ]  Multi frame onion skin.
 * [ ]  BUG: Changing the resolution sometimes causes crash.
 * [ ]  Support DSLR camera.
 * [ ]  Add "No Camera" item to the camera combo box.
 * [ ]  Add a "Mirror" option.

Some of these are addressed tangentially in #2635 Stop Motion Support by @turtletooth

**Comment by quirinux-so** [comment]
In OpenToonz 1.3 installed in Debian 10 from Debian Multimedia my webcam works fine, but in version 1.4 it is only detected (although it doesn't work). To use version 1.4 I compiled from the code since none of the existing versions in snap or appimage detect my webcam (nor other webcams that I tried). I leave more information about my incident, with screenshots, at:
https://github.com/opentoonz/opentoonz/issues/3333

**Comment by quirinux-so** [comment]
I have compiled in Debian, from sources, version 1.3, 1.4 and nightly build version 2020-06-18 and only in 1.3 does the webcam work.

In the other versions, the webcam is detected but does not work. I have the impression that perhaps it can be solved by making a small change. 

Please review it, since I intend to include OpenToonz 1.4 (or newer) in the next version to come out of Quirinux (www.quirinux.org) a GNU / Linux distribution from Argentina and Spain, that will seek to standardize free software in the area of animation. 

A greeting.

:-)

Charlie
cmartinez@quirinux.org

**Comment by RodneyBaker** [comment]
@shun-iwasawa  As you find time...

It is not clear how the recent addition of stopmotion DSLR and web camera support has effected the goals for the basic Camera Capture here as a work in progress.

My initially take is that both features are highly useful but in different ways.
Perhaps one way forward might be to round this feature set off and document it well as a basic line tester and reference the other camera features for extended camera and stop motion use.  This especially as the latter seems to implement many of the goals set forward here.

If this was the approach then future work would streamline the vision for both with the goal of keeping the basic line tester as simple as possible while making stopmotion a larger extensible feature set.

There is also the element that not all cameras are supported by the larger feature set of Canon SDK.
So much of the work ahead might be in the form of simply community documentation of what we have so users get a clear view on what is available.

Over to you...

@quirinux-so (Charlie) Much time has passed since your last comment.  It will be good if you can give us a status update.  
Thanks!

**Comment by quirinux-so** [comment]
> @shun-iwasawa As you find time...
> 
> It is not clear how the recent addition of stopmotion DSLR and web camera support has effected the goals for the basic Camera Capture here as a work in progress.
> 
> My initially take is that both features are highly useful but in different ways.
> Perhaps one way forward might be to round this feature set off and document it well as a basic line tester and reference the other camera features for extended camera and stop motion use. This especially as the latter seems to implement many of the goals set forward here.
> 
> If this was the approach then future work would streamline the vision for both with the goal of keeping the basic line tester as simple as possible while making stopmotion a larger extensible feature set.
> 
> There is also the element that not all cameras are supported by the larger feature set of Canon SDK.
> So much of the work ahead might be in the form of simply community documentation of what we have so users get a clear view on what is available.
> 
> Over to you...
> 
> @quirinux-so (Charlie) Much time has passed since your last comment. It will be good if you can give us a status update.
> Thanks!

Thanks Rodney:

For stop-motion I am using TupiTube Desk, its developer (one only person) achieved compatibility with webcam and DSLR cameras through some libraries, maybe you could use your project as a reference: 

https://github.com/xtingray/tupitube.desk

I don't have a camera DSLR, anyway the tab for Stop Motion was not built into GNU / Linux, which I use. I miss working with StopMotion in OpenToonz using a webcam, but although my last comment is old nothing has changed: whether I use snap, flatpak, appimage, or compile from source to version 1.4 or the Nigthly Builds, I still cannot use my camera. The only way Open Toonz allows me to use it is by adding the multimedia repositories and installing it from there, but only in version 1.3 ... I keep wondering what changed. It also doesn't work in Morevna.

I have been told that in the OTX version the webcam works.

A greeting.

**Comment by manongjohn** [comment]
@quirinux-so You might want to check out Tahoma, an OT variant focused on 2D Digital Animation and Stop Motion. A Linux version (webcam support only, no Canon support) was recently made available.

**Comment by quirinux-so** [comment]
@manongjohn Thank you very much, after receiving and answering the email just, in my inbox the following was the notice that you left me on YouTube about this appimage. I'm very excited, because it actually seems very stable and fast on Debian. The camera issue is almost resolved, it would still be necessary to resolve something:

https://github.com/turtletooth/tahoma/issues/250

**Comment by motoglaf** [comment]
Camera capture is really the only pencil-testing setup I use. The only feature I would suggest adding is a feedback noise after every shot. This is in the stop motion panel but that way does not work with my setup. I would imagine this would not be a hard thing to add as it's already in another part of the program. It could be a click or a beep just something to signal that the shot was completed so you don't have to reference your monitor after every shot. It would be very useful especially if you use an external number pad or controller to take the shots not near the computer.

**Comment by RodneyBaker** [comment]
@shun-iwasawa 
I have been moving proposals, outlines and such postings to Discussions - Roadmap.
If this is inconvenient please let me know and I'll gladly transfer it back.
